subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -188,25 +212,64 @@ private String getFilename(Path dir) {
 
         @Override
         public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
+            if (attrs.isSymbolicLink()) {
+                BasicFileAttributes targetAttributes = readAttributesOfSymlinkTarget(file, attrs);
+                if (targetAttributes.isDirectory()) {
+                    try {
+                        // TODO: Detect cycles
+                        Path targetDir = file.toRealPath();
+                        symbolicLinkMappings.addFirst(new SymbolicLinkMapping(file.toString(), targetDir.toString()));
+                        Files.walkFileTree(targetDir, EnumSet.noneOf(FileVisitOption.class), Integer.MAX_VALUE, this);
+                        symbolicLinkMappings.removeFirst();
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(String.format("Could not list contents of directory '%s'.", file), e);
+                    }
+                } else {
+                    visitResolvedFile(file, targetAttributes, AccessType.VIA_SYMLINK);
+                }
+            } else {
+                visitResolvedFile(file, attrs, AccessType.DIRECT);
+            }
+            return FileVisitResult.CONTINUE;
+        }
+
+        private String remapAbsolutePath(Path dir) {
+            String targetAbsolutePath = dir.toString();
+            return symbolicLinkMappings.stream()
+                .map(mapping -> mapping.remapPath(targetAbsolutePath))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .findFirst()
+                .orElse(targetAbsolutePath);
+        }
+
+        private void visitResolvedFile(Path file, BasicFileAttributes targetAttributes, AccessType accessType) {
             String internedName = intern(file.getFileName().toString());
             if (shouldVisit(file, internedName, false, builder.getRelativePath())) {
-                builder.visitFile(snapshotFile(file, internedName, attrs));
+                builder.visitFile(snapshotFile(file, internedName, targetAttributes, accessType));
+            }
+        }
+
+        private BasicFileAttributes readAttributesOfSymlinkTarget(Path symlink, BasicFileAttributes symlinkAttributes) {
+            try {
+                return Files.readAttributes(symlink, BasicFileAttributes.class);
+            } catch (IOException ioe) {
```

Comment:
We should only catch `NoSuchFileException` here.

Also, it's unclear why we return the symlink's attributes when the link is dangling. Can you clarify that?

Reply:
We should catch all `IOException`s here as we do for `stat` and as we did before.

We return the symlinks attributes since that is the behaviour of `walkFileTree` when following symlinks.

Reply:
> We return the symlinks attributes since that is the behaviour of `walkFileTree` when following symlinks.

You mean `Files.walkFileTree()`? I think this would be worth explaining in a comment here.

Reply:
I added a comment.